blocks/monitoring/__init__.py

```diff
@@ -0,0 +1,193 @@
+"""This part of the framework helps you monitor your model during training.
+
+:class:`MonitorChannels` describe quantities that should be monitored. They
+can either be Theano variables (e.g. the objective function, the weight
+norms, etc.) or they can be callable functions (e.g. if you want to use
+NumPy to calculate eigenvalues, or sample text from your language model).
+
+.. todo::
+
+   Monitored Theano variables might depend on inputs beyond those given to
+   the model. There should be a way of providing these to the monitoring
+   channels, collecting them, and compiling them as part of the Theano
+   function that performs monitoring.
+
+"""
+import logging
+from abc import ABCMeta, abstractmethod
+
+from blocks.utils import shared_expression, update_instance
+
+logger = logging.getLogger(__name__)
+
+
+class MonitorChannel(object):
```

Comment:
On 1/7/2015 6:27 PM, Bart van Merriënboer wrote:
>
> In blocks/monitoring/__init__.py 
> <https://github.com/bartvm/blocks/pull/90#discussion-diff-22600412>:
>
> > +
> > +   Monitored Theano variables might depend on inputs beyond those given to
> > +   the model. There should be a way of providing these to the monitoring
> > +   channels, collecting them, and compiling them as part of the Theano
> > +   function that performs monitoring.
> > +
> > +"""
> > +import logging
> > +from abc import ABCMeta, abstractmethod
> > +
> > +from blocks.utils import shared_expression, update_instance
> > +
> > +logger = logging.getLogger(__name__)
> > +
> > +
> > +class MonitorChannel(object):
>
> I don't understand what you mean with having many different sources 
> with the same name? If you mean across different datasets, then it is 
> true that the source is always called |features|. However, I don't 
> think that is much of a problem. We can't guarantee global uniqueness 
> anyway, so if people want to combine multiple datasets they might as 
> well prefix the dataset name in front of the source.
>
In PL2 you can ask the dataset to provide the same source in different 
spaces. Say you ask for tragets in 1-hot and integer representation (or 
however these are called in PL2). Both are called "features", though. 
Thus you cannot return the batch as a dict. I spoke once about it with 
Pascal Lamblin.
>
> The ability to ask a dataset for specific sources only might be worth 
> considering. It would be a shame if the user is forced to read large 
> amounts of data from disk just to reject them afterwards (e.g. if a 
> dataset consists of images, tags, and a category, and we only want the 
> tags and category).
>
Yes, this ability is useful. I don't know whether it should be automatic 
(as in PL2, or we walk the graph, find the inputs, and ask for their 
names), or manual.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/90/files#r22600412>.
>